src/lib.rs

```diff
@@ -0,0 +1,1039 @@
+//! A concurrent work-stealing deque.
+//!
+//! The data structure can be thought of as a dynamically growable and shrinkable buffer that has
+//! two ends: bottom and top. A [`Deque`] can [`push`] elements into the bottom and [`pop`]
+//! elements from the bottom, but it can only [`steal`][Deque::steal] elements from the top.
+//!
+//! A [`Deque`] doesn't implement `Sync` so it cannot be shared among multiple threads. However, it
+//! can create [`Stealer`]s, and those can be easily cloned, shared, and sent to other threads.
+//! [`Stealer`]s can only [`steal`][Stealer::steal] elements from the top.
+//!
+//! Here's a visualization of the data structure:
+//!
+//! ```text
+//!                    top
+//!                     _
+//!    Deque::steal -> | | <- Stealer::steal
+//!                    | |
+//!                    | |
+//!                    | |
+//! Deque::push/pop -> |_|
+//!
+//!                  bottom
+//! ```
+//!
+//! # Work-stealing schedulers
+//!
+//! Usually, the data structure is used in work-stealing schedulers as follows.
+//!
+//! There is a number of threads. Each thread owns a [`Deque`] and creates a [`Stealer`] that is
+//! shared among all other threads. Alternatively, it creates multiple [`Stealer`]s - one for each
+//! of the other threads.
+//!
+//! Then, all threads are executing in a loop. In the loop, each one attempts to [`pop`] some work
+//! from its own [`Deque`]. But if it is empty, it attempts to [`steal`][Stealer::steal] work from
+//! some other thread instead. When executing work (or being idle), a thread may produce more work,
+//! which gets [`push`]ed into its [`Deque`].
+//!
+//! Of course, there are many variations of this strategy. For example, sometimes it may be
+//! beneficial for a thread to always [`steal`][Deque::steal] work from the top of its deque
+//! instead of calling [`pop`] and taking it from the bottom.
```

Comment:
I'm just curious, what would be the benefit of using `fn Deque::steal()` instead of `fn Deque::pop()`?

Reply:
Rayon uses `steal` instead of `pop` when in the "breadth-first" mode, which gives a pretty large speedup in Stylo. See rayon-rs/rayon#360 for more details.